flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformerITCase.java

```diff
@@ -0,0 +1,165 @@
+                    handler.addOrUpdateFuture.get(2000, TimeUnit.MILLISECONDS);
+                }
+                client.deleteConfigMapsByLabels(labels).get();
+                for (TestingCallbackHandler handler : callbackHandlers) {
+                    handler.deleteFuture.get(2000, TimeUnit.MILLISECONDS);
+                }
+            } finally {
+                client.deleteConfigMapsByLabels(labels).get();
+            }
+            watchers.forEach(Watch::close);
+        }
+    }
+
+    private void createConfigMapsInRange(
+            FlinkKubeClient client,
+            Map<String, String> labels,
+            String namePrefix,
+            int start,
+            int end)
+            throws Exception {
+        for (int i = start; i < end; i++) {
+            client.createConfigMap(
+                            new KubernetesConfigMap(
+                                    new ConfigMapBuilder()
+                                            .withNewMetadata()
+                                            .withName(namePrefix + i)
+                                            .withLabels(labels)
+                                            .endMetadata()
+                                            .build()))
+                    .get();
+        }
+    }
+
+    private static final class TestingCallbackHandler
+            extends NoOpWatchCallbackHandler<KubernetesConfigMap> {
+
+        private final CompletableFuture<Void> addFuture = new CompletableFuture<>();
+        private final CompletableFuture<Void> addOrUpdateFuture = new CompletableFuture<>();
+        private final CompletableFuture<Void> deleteFuture = new CompletableFuture<>();
+
+        private final String name;
+
+        private TestingCallbackHandler(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public void onAdded(List<KubernetesConfigMap> resources) {
+            final KubernetesConfigMap kubernetesConfigMap = resources.get(0);
+            assertThat(kubernetesConfigMap.getName(), is(name));
+            addFuture.complete(null);
+            final String foo = kubernetesConfigMap.getData().get("foo");
+            if (foo != null) {
+                assertThat(foo, is("bar"));
+                addOrUpdateFuture.complete(null);
+            }
+        }
+
+        @Override
+        public void onModified(List<KubernetesConfigMap> resources) {
+            final KubernetesConfigMap kubernetesConfigMap = resources.get(0);
+            assertThat(kubernetesConfigMap.getName(), is(name));
+            final String foo = kubernetesConfigMap.getData().get("foo");
+            assertThat(foo, is("bar"));
+            if (addOrUpdateFuture.isDone()) {
+                assertNotNull(
+                        kubernetesConfigMap
+                                .getInternalResource()
+                                .getMetadata()
+                                .getDeletionTimestamp());
+            } else {
```

Comment:
Could we guarantee that the `onAdded` always happens before `onModified`?

Reply:
Yeah, i will add a validation for it.